storage/src/test/java/org/apache/kafka/storage/internals/log/LogSegmentTest.java

```diff
@@ -0,0 +1,763 @@
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.HashMap;
+import java.util.Iterator;
+import java.util.List;
+import java.util.Map;
+import java.util.Optional;
+import java.util.OptionalLong;
+
+import static org.junit.jupiter.api.Assertions.assertEquals;
+import static org.junit.jupiter.api.Assertions.assertFalse;
+import static org.junit.jupiter.api.Assertions.assertNull;
+import static org.junit.jupiter.api.Assertions.assertThrows;
+import static org.junit.jupiter.api.Assertions.assertTrue;
+
+public class LogSegmentTest {
+    private final TopicPartition topicPartition = new TopicPartition("topic", 0);
+    private final List<LogSegment> segments = new ArrayList<>();
+    private File logDir = null;
+
+    /* Create a segment with the given base offset */
+    private LogSegment createSegment(long offset, int indexIntervalBytes, Time time) throws IOException {
+        LogSegment seg = LogTestUtils.createSegment(offset, logDir, indexIntervalBytes, time);
+        segments.add(seg);
+        return seg;
+    }
+
+    private LogSegment createSegment(long offset) throws IOException {
+        return createSegment(offset, 10, Time.SYSTEM);
+    }
+
+    private LogSegment createSegment(long offset, Time time) throws IOException {
+        return createSegment(offset, 10, time);
+    }
+
+    private LogSegment createSegment(long offset, int indexIntervalBytes) throws IOException {
+        return createSegment(offset, indexIntervalBytes, Time.SYSTEM);
+    }
+
+    /* Create a ByteBufferMessageSet for the given messages starting from the given offset */
+    private MemoryRecords records(long offset, String... records) {
+        List<SimpleRecord> simpleRecords = new ArrayList<>();
+        for (String s : records) {
+            simpleRecords.add(new SimpleRecord(offset * 10, s.getBytes()));
+        }
+        return MemoryRecords.withRecords(
+            RecordBatch.MAGIC_VALUE_V1, offset,
+            Compression.NONE, TimestampType.CREATE_TIME, simpleRecords.toArray(new SimpleRecord[0]));
+    }
+
+    @BeforeEach
+    public void setup() {
+        logDir = TestUtils.tempDirectory();
+    }
+
+    @AfterEach
+    public void teardown() throws IOException {
+        for (LogSegment segment : segments) {
+            segment.close();
+        }
+        Utils.delete(logDir);
+    }
+
+    /**
+     * LogSegmentOffsetOverflowException should be thrown while appending the logs if:
+     * 1. largestOffset - baseOffset < 0
+     * 2. largestOffset - baseOffset > Integer.MAX_VALUE
+     */
+    @ParameterizedTest
+    @CsvSource({
+        "0, -2147483648",
+        "0, 2147483648",
+        "1, 0",
+        "100, 10",
+        "2147483648, 0",
+        "-2147483648, 0",
+        "2147483648,4294967296"
+    })
+    public void testAppendForLogSegmentOffsetOverflowException(long baseOffset, long largestOffset) throws IOException {
+        LogSegment seg = createSegment(baseOffset, 10, Time.SYSTEM);
```

Comment:
nit: It seems that in the Scala version, `LogSegment` wasn't being closed, but I think we could leverage the `try-with-resources` statement to make the test more stable and avoid potential issues.

Reply:
Please also apply to the following test if needed.